capa/ghidra/capa_explorer.py

```diff
@@ -0,0 +1,261 @@
+    try:
+        selected_dir = askDirectory("Choose capa rules directory", "Ok")  # type: ignore [name-defined] # noqa: F821
+        if selected_dir:
+            rules_dir = selected_dir.getPath()
+    except RuntimeError:
+        # RuntimeError thrown when user selects "Cancel"
+        pass
+
+    if not rules_dir:
+        logger.info("You must choose a capa rules directory before running capa.")
+        return capa.main.E_MISSING_RULES
+
+    rules_path: pathlib.Path = pathlib.Path(rules_dir)
+    logger.info("running capa using rules from %s", str(rules_path))
+
+    rules = capa.main.get_rules([rules_path])
+    meta = capa.ghidra.helpers.collect_metadata([rules_path])
+    extractor = capa.features.extractors.ghidra.extractor.GhidraFeatureExtractor()
+
+    capabilities, counts = capa.main.find_capabilities(rules, extractor, True)
+
+    if capa.main.has_file_limitation(rules, capabilities, is_standalone=False):
+        logger.info("capa encountered warnings during analysis")
+
+    return capa.render.json.render(meta, rules, capabilities)
+
+
+def get_locations(match_dict):
+    """recursively collect match addresses and associated nodes"""
+    if "locations" in match_dict:
+        for loc in match_dict.get("locations", {}):
+            if "type" in loc:
+                # either an rva (absolute)
+                # or an offset into a file (file)
+                if loc.get("type", "") in ("absolute", "file"):
+                    yield loc.get("value"), match_dict.get("node")
+
+    if match_dict["children"]:
+        for child in match_dict.get("children", {}):
+            yield from get_locations(child)
+
+
+def parse_json(capa_data):
+    """Parse json produced by capa"""
+    ghidra_data = []
+
+    rules = capa_data["rules"]
+    for rule in rules.keys():
+        # loosely coupled location & node data lists
+        this_locs = []
+        this_node: List[Dict] = []
+
+        # dict data of currently matched rule
+        this_capability = rules[rule]
+        meta = this_capability["meta"]
+
+        # get Mitre ATT&CK and MBC
+        this_attack = meta.get("attack")
+        this_mbc = meta.get("mbc")
+
+        # scope match for the rule
+        this_scope = meta["scopes"].get("static")
+        with suppress(KeyError):
+            # always grab first location of match
+            this_locs.append(this_capability["matches"][0][0]["value"])
+            # align node data
+            this_node.append({})
+
+        if "namespace" in meta:
+            # split into list to help define child namespaces
+            # in ghidra
+            namespace_str = Namespace.DELIMITER.join(meta["namespace"].split("/"))
+            this_namespace = "capa" + Namespace.DELIMITER + namespace_str
+        else:
+            this_namespace = "capa"
+
+        # recurse to find all locations
+        # grab second dict, containing additional matches
+        # and node data
+        matches = this_capability["matches"][0][1]
```

Comment:
What happens if `matches` is empty?

Reply:
I may be wrong, but if there were no matches, then it wouldn't be included in the `capa_data` at all. So I guess the bigger missed check is handling if a sample contains 0 matches for any rules.

Reply:
As a follow-up, I've found that if a binary contains no matches, the generated json will include `capa_data['meta']`, but the `'rules'` key will contain no values. This would cause the entire `parse_json()` function to be skipped.

I've added a popup to notify the user that no matches were found.